src/data_source/storage/sql.rs

```diff
@@ -515,19 +546,39 @@ where
         Ok(rows.map(|res| parse_block(res?)).collect().await)
     }
 
+    async fn get_payload_range<R>(
+        &self,
+        range: R,
+    ) -> QueryResult<Vec<QueryResult<PayloadQueryData<Types>>>>
+    where
+        R: RangeBounds<usize> + Send,
+    {
+        let (where_clause, params) = bounds_to_where_clause(range, "h.height");
+        let query = format!(
+            "SELECT {PAYLOAD_COLUMNS}
+              FROM header AS h
+              JOIN payload AS p ON h.height = p.height
+              {where_clause}
+              ORDER BY h.height ASC"
+        );
+        let rows = self.query(&query, params).await?;
+
+        Ok(rows.map(|res| parse_payload(res?)).collect().await)
+    }
+
     async fn get_block_with_transaction(
         &self,
         hash: TransactionHash<Types>,
     ) -> QueryResult<(BlockQueryData<Types>, TransactionIndex<Types>)> {
-        // ORDER BY t.id ASC ensures that if there are duplicate transactions, we return the first
+        // ORDER BY ASC ensures that if there are duplicate transactions, we return the first
         // one.
         let query = format!(
             "SELECT {BLOCK_COLUMNS}, t.index AS tx_index
                 FROM header AS h
                 JOIN payload AS p ON h.height = p.height
                 JOIN transaction AS t ON t.block_height = h.height
                 WHERE t.hash = $1
-                ORDER BY t.id ASC
+                ORDER BY (t.block_height, t.index) ASC
```

Comment:
Now we see the importance of this schema change: this deterministic primary key gives us the first transaction in chronological order, whereas the old serial column would just order by when this particular DB first saw each transaction (possibly out of order)